homeassistant/components/google_travel_time/sensor.py

```diff
@@ -133,62 +96,85 @@ def convert_time_to_utc(timestr):
     return dt_util.as_timestamp(combined)
 
 
-def setup_platform(hass, config, add_entities_callback, discovery_info=None):
-    """Set up the Google travel time platform."""
-
-    def run_setup(event):
-        """
-        Delay the setup until Home Assistant is fully initialized.
-
-        This allows any entities to be created already
-        """
-        hass.data.setdefault(DATA_KEY, [])
-        options = config.get(CONF_OPTIONS)
-
-        if options.get("units") is None:
-            options["units"] = hass.config.units.name
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: Callable[[list[SensorEntity], bool], None],
+) -> None:
+    """Set up a Google travel time sensor entry."""
+    name = None
+    if not config_entry.options:
+        new_data = config_entry.data.copy()
+        options = new_data.pop(CONF_OPTIONS, {})
+        name = new_data.pop(CONF_NAME, None)
 
-        travel_mode = config.get(CONF_TRAVEL_MODE)
-        mode = options.get(CONF_MODE)
+        if CONF_UNITS not in options:
+            options[CONF_UNITS] = hass.config.units.name
 
-        if travel_mode is not None:
+        if CONF_TRAVEL_MODE in new_data:
             wstr = (
                 "Google Travel Time: travel_mode is deprecated, please "
                 "add mode to the options dictionary instead!"
             )
             _LOGGER.warning(wstr)
-            if mode is None:
+            travel_mode = new_data.pop(CONF_TRAVEL_MODE)
+            if CONF_MODE not in options:
                 options[CONF_MODE] = travel_mode
 
-        titled_mode = options.get(CONF_MODE).title()
-        formatted_name = f"{DEFAULT_NAME} - {titled_mode}"
-        name = config.get(CONF_NAME, formatted_name)
-        api_key = config.get(CONF_API_KEY)
-        origin = config.get(CONF_ORIGIN)
-        destination = config.get(CONF_DESTINATION)
+        if CONF_MODE not in options:
+            options[CONF_MODE] = "driving"
 
-        sensor = GoogleTravelTimeSensor(
-            hass, name, api_key, origin, destination, options
+        await hass.config_entries.async_update_entry(
+            config_entry, data=new_data, options=options
         )
-        hass.data[DATA_KEY].append(sensor)
 
-        if sensor.valid_api_connection:
-            add_entities_callback([sensor])
+    hass.data.setdefault(DATA_KEY, [])
+
+    api_key = config_entry.data.get(CONF_API_KEY)
+    origin = config_entry.data.get(CONF_ORIGIN)
+    destination = config_entry.data.get(CONF_DESTINATION)
+    name = name or f"{DEFAULT_NAME}: {origin} -> {destination}"
+
+    sensor = GoogleTravelTimeSensor(
```

Comment:
There are side effects in the init method of the entity. We need to move that out of the init method and schedule it on the executor before creating the entity.

Reply:
actually I'm not sure that we even need to call the update function because we already do that in `async_added_to_hass`